IllegalStateException during indexing

I got the next exception from Jandex library:

Caused by: java.lang.IllegalStateException: Class not found in class table:scala.Function18$$anonfun$curried$1
    at org.jboss.jandex.IndexWriterV2.positionOf (IndexWriterV2.java:507)
    at org.jboss.jandex.IndexWriterV2.writeTypeEntry (IndexWriterV2.java:789)
    at org.jboss.jandex.IndexWriterV2.writeTypeTable (IndexWriterV2.java:271)
    at org.jboss.jandex.IndexWriterV2.write (IndexWriterV2.java:211)
    at org.jboss.jandex.IndexWriter.write (IndexWriter.java:110)
    at org.jboss.jandex.IndexWriter.write (IndexWriter.java:74)
    at org.jboss.jandex.JarIndexer.createJarIndex (JarIndexer.java:193)

Please investigate this issue. thanks

--- a/core/src/main/java/org/jboss/jandex/DotName.java
+++ b/core/src/main/java/org/jboss/jandex/DotName.java
@@ -347,7 +347,7 @@ public final class DotName implements Comparable<DotName> {
         if (other.componentized && !componentized)
             return crossEquals(this, other);
 
-        return prefix != null && innerClass == other.innerClass && local.equals(other.local) && prefix.equals(other.prefix);
+        return componentizedEquals(this, other);
     }
 
     private static boolean crossEquals(final DotName simple, final DotName comp) {
@@ -385,6 +385,66 @@ public final class DotName implements Comparable<DotName> {
         }
         //And finally, verify we consumed it all:
         return cursor == -1;
+    }
+
+    private static boolean componentizedEquals(DotName a, DotName b) {
+        // fast path for identical componentizations
+        if (a == null && b == null) {
+            return true;
+        }
+        if (a == null || b == null) {
+            return false;
+        }
+        if (a.innerClass == b.innerClass && a.local.equals(b.local)) {
+            return componentizedEquals(a.prefix, b.prefix);
+        }
+
+        // this algorithm simply goes from the end towards the beginning, both `DotName`s in parallel,
+        // and checks that they match on each position; whenever there is a mismatch, the result is `false`
+        //
+        // positions range from -1 to <local name length - 1>, where values >= 0 are indices into
+        // the string and the -1 value is used for a separator (either '$' or '.', depending on
+        // whether given name has the `innerClass` flag set)
+        //
+        // an interesting situation occurs at the position -1 of the farthest `DotName` (one that has
+        // no prefix): we still need to compare whether the `innerClass` flags match, but there's no need
+        // to write a special case for it, we just treat a prefix-less `DotName` that is _not_ flagged `innerClass`
+        // as having an extra '.' character at the beginning
+        //
+        // the same algorithm could also be used for `crossEquals`, but that method is supposedly more efficient
+        // for its special case
+
+        String aLocal = a.local;
+        String bLocal = b.local;
+        int aPos = aLocal.length() - 1;
+        int bPos = bLocal.length() - 1;
+        while (a != null && b != null) {
+            char aChar = aPos >= 0 ? aLocal.charAt(aPos) : (a.innerClass ? '$' : '.');
+            char bChar = bPos >= 0 ? bLocal.charAt(bPos) : (b.innerClass ? '$' : '.');
+
+            if (aChar != bChar) {
+                return false;
+            }
+
+            aPos--;
+            if (aPos < -1) {
+                a = a.prefix;
+                if (a != null) {
+                    aLocal = a.local;
+                    aPos = aLocal.length() - 1;
+                }
+            }
+
+            bPos--;
+            if (bPos < -1) {
+                b = b.prefix;
+                if (b != null) {
+                    bLocal = b.local;
+                    bPos = bLocal.length() - 1;
+                }
+            }
+        }
+        return a == null && b == null;
     }
 
     private static class IndexState {

--- a/core/src/main/java/org/jboss/jandex/IndexReaderV2.java
+++ b/core/src/main/java/org/jboss/jandex/IndexReaderV2.java
@@ -181,21 +181,33 @@ final class IndexReaderV2 extends IndexReaderImpl {
 
         nameTable = new DotName[entries];
         for (int i = 1; i < entries; i++) {
-            int depth = stream.readPackedU32();
-            boolean inner = (depth & 1) == 1;
-            depth >>= 1;
+            // see IndexWriterV2.writeNameTable
+            if (version >= 11) {
+                int prefixOffset = stream.readPackedU32();
+                boolean inner = (prefixOffset & 1) == 1;
+                prefixOffset >>= 1;
 
-            String local = stringTable[stream.readPackedU32()];
+                int prefixPosition = prefixOffset == 0 ? 0 : i - prefixOffset;
+                DotName prefix = nameTable[prefixPosition];
+                String local = stringTable[stream.readPackedU32()];
+                nameTable[i] = new DotName(prefix, local, true, inner);
+            } else {
+                int depth = stream.readPackedU32();
+                boolean inner = (depth & 1) == 1;
+                depth >>= 1;
 
-            if (depth <= lastDepth) {
-                while (lastDepth-- >= depth) {
-                    assert curr != null;
-                    curr = curr.prefix();
+                String local = stringTable[stream.readPackedU32()];
+
+                if (depth <= lastDepth) {
+                    while (lastDepth-- >= depth) {
+                        assert curr != null;
+                        curr = curr.prefix();
+                    }
                 }
-            }
 
-            nameTable[i] = curr = new DotName(curr, local, true, inner);
-            lastDepth = depth;
+                nameTable[i] = curr = new DotName(curr, local, true, inner);
+                lastDepth = depth;
+            }
         }
     }
 

--- a/core/src/main/java/org/jboss/jandex/IndexWriterV2.java
+++ b/core/src/main/java/org/jboss/jandex/IndexWriterV2.java
@@ -429,18 +429,35 @@ final class IndexWriterV2 extends IndexWriterImpl {
         // Zero is reserved for null
         int pos = 1;
         for (Entry<String, DotName> entry : sortedNameTable.entrySet()) {
-            nameTable.put(entry.getValue(), pos++);
+            nameTable.put(entry.getValue(), pos);
             DotName name = entry.getValue();
             assert name.isComponentized();
 
-            int nameDepth = 0;
-            for (DotName prefix = name.prefix(); prefix != null; prefix = prefix.prefix())
-                nameDepth++;
+            if (version >= 11) {
+                // to save space, instead of storing [absolute] prefix position, we store [relative] prefix offset,
+                // and since the offset is always negative (the prefix must have been written before),
+                // we store its absolute value; the value of 0 is used to mean `null` prefix
+                int prefixPosition = name.prefix() == null ? 0 : positionOf(name.prefix());
+                int prefixOffset = prefixPosition == 0 ? 0 : pos - prefixPosition;
 
-            nameDepth = nameDepth << 1 | (name.isInner() ? 1 : 0);
+                int prefixOffsetToWrite = prefixOffset << 1 | (name.isInner() ? 1 : 0);
+                stream.writePackedU32(prefixOffsetToWrite);
+                stream.writePackedU32(positionOf(name.local()));
+            } else {
+                // in older versions, we store the depth of the name, in hope that we can find the prefix
+                // when reading by "unrolling" the previously read name; that mostly works, but may fail
+                // in case of weird names starting with '$'
+                int nameDepth = 0;
+                for (DotName prefix = name.prefix(); prefix != null; prefix = prefix.prefix())
+                    nameDepth++;
 
-            stream.writePackedU32(nameDepth);
-            stream.writePackedU32(positionOf(name.local()));
+                nameDepth = nameDepth << 1 | (name.isInner() ? 1 : 0);
+
+                stream.writePackedU32(nameDepth);
+                stream.writePackedU32(positionOf(name.local()));
+            }
+
+            pos++;
         }
     }
 
@@ -488,7 +505,7 @@ final class IndexWriterV2 extends IndexWriterImpl {
     private int positionOf(DotName className) {
         Integer i = nameTable.get(className);
         if (i == null)
-            throw new IllegalStateException("Class not found in class table:" + className);
+            throw new IllegalStateException("Class not found in class table: " + className);
 
         return i.intValue();
     }

--- a/core/src/main/java/org/jboss/jandex/Type.java
+++ b/core/src/main/java/org/jboss/jandex/Type.java
@@ -116,13 +116,13 @@ public abstract class Type {
     }
 
     /**
-     * Creates a type instance of the specified kind. Types of kind <code>CLASS</code>,
-     * directly use the specified name. Types of kind <code>ARRAY</code> parse the name
-     * in the Java reflection format (Java descriptor format changing / to '.',
-     * e.g. "[[[[Ljava.lang.String;"). Types of kind PRIMITIVE parsed using the
-     * primitive descriptor format (e.g. "I" for int).
-     * Types of kind VOID ignore the specified name, and return a void type. All
-     * other types will throw an exception.
+     * Creates a type instance of the specified kind. If {@code kind} is {@code CLASS},
+     * the {@code name} is used as is. If {@code kind} is {@code ARRAY}, the {@code name}
+     * must be in the Java reflection format (Java descriptor format changing {@code /}
+     * to {@code .}, e.g. {@code [[[[Ljava.lang.String;"} or {@code [[[I}). If {@code kind}
+     * is {@code PRIMITIVE}, the name must be in the Java reflection format (which is equal
+     * to the Java keyword denoting the primitive type, e.g. {@code int}). If kind is
+     * {@code VOID}, the {@code name} is ignored. All other kinds cause an exception.
      *
      * @param name the name to use or parse
      * @param kind the kind of type to create

--- a/core/src/test/java/org/jboss/jandex/test/DotNameComponentizationTest.java
+++ b/core/src/test/java/org/jboss/jandex/test/DotNameComponentizationTest.java
@@ -6,6 +6,7 @@ import java.io.IOException;
 
 import org.jboss.jandex.DotName;
 import org.jboss.jandex.Index;
+import org.jboss.jandex.test.util.IndexingUtil;
 import org.junit.jupiter.api.Test;
 
 public class DotNameComponentizationTest {
@@ -23,6 +24,8 @@ public class DotNameComponentizationTest {
         Index index = Index.of(Foo.class);
 
         doTest(index);
+
+        doTest(IndexingUtil.roundtrip(index));
     }
 
     private void doTest(Index index) {

--- a/core/src/test/java/org/jboss/jandex/test/DotNameComponentizationTest.java
+++ b/core/src/test/java/org/jboss/jandex/test/DotNameComponentizationTest.java
@@ -1,0 +1,55 @@
+package org.jboss.jandex.test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import java.io.IOException;
+
+import org.jboss.jandex.DotName;
+import org.jboss.jandex.Index;
+import org.junit.jupiter.api.Test;
+
+public class DotNameComponentizationTest {
+    static class Foo<T> {
+        // must not be `static` in this test
+        class $Bar {
+        }
+
+        static <A> void baz(Foo<A>.$Bar foobar) {
+        }
+    }
+
+    @Test
+    public void test() throws IOException {
+        Index index = Index.of(Foo.class);
+
+        doTest(index);
+    }
+
+    private void doTest(Index index) {
+        assertEquals(1, index.getKnownClasses().size());
+        DotName name = index.getKnownClasses().iterator().next().firstMethod("baz").parameters().get(0).name();
+
+        {
+            DotName org = DotName.createComponentized(null, "org");
+            DotName jboss = DotName.createComponentized(org, "jboss");
+            DotName jandex = DotName.createComponentized(jboss, "jandex");
+            DotName test = DotName.createComponentized(jandex, "test");
+            DotName dotNameComponentizationTest = DotName.createComponentized(test, "DotNameComponentizationTest");
+            DotName foo = DotName.createComponentized(dotNameComponentizationTest, "Foo", true);
+            DotName bar = DotName.createComponentized(foo, "$Bar", true);
+
+            assertEquals(bar, name);
+        }
+
+        {
+            DotName org = DotName.createComponentized(null, "org");
+            DotName jboss = DotName.createComponentized(org, "jboss");
+            DotName jandex = DotName.createComponentized(jboss, "jandex");
+            DotName test = DotName.createComponentized(jandex, "test");
+            DotName dotNameComponentizationTest = DotName.createComponentized(test, "DotNameComponentizationTest");
+            DotName fooBar = DotName.createComponentized(dotNameComponentizationTest, "Foo$$Bar", true);
+
+            assertEquals(fooBar, name);
+        }
+    }
+}

--- a/core/src/test/java/org/jboss/jandex/test/DotNameTestCase.java
+++ b/core/src/test/java/org/jboss/jandex/test/DotNameTestCase.java
@@ -84,6 +84,84 @@ public class DotNameTestCase {
         }
     }
 
+    private static void definitelyEquals(DotName a, DotName b) {
+        assertEquals(a, b);
+        assertEquals(b, a);
+        assertEquals(a.compareTo(b), 0);
+        assertEquals(b.compareTo(a), 0);
+        sameHashCode(a, b);
+    }
+
+    private static void sameHashCode(DotName a, DotName b) {
+        assertEquals(a.hashCode(), b.hashCode());
+    }
+
+    private static void definitelyNotEquals(DotName a, DotName b) {
+        assertFalse(a.equals(b), "should not be equals");
+        assertFalse(b.equals(a), "should not be equals");
+        assertFalse(a.compareTo(b) == 0);
+        assertFalse(b.compareTo(a) == 0);
+        assertEquals(b.compareTo(a), -1 * a.compareTo(b));
+    }
+
+    private static DotName createRandomDotName() {
+        return r.nextBoolean() ? createRandomComponentised() : createRandomSimple();
+    }
+
+    private static DotName createRandomSimple() {
+        int partsToAppend = r.nextInt(5);
+        final StringBuilder sb = new StringBuilder();
+        sb.append(someRandomChars());
+        while (partsToAppend > 0) {
+            sb.append('.');
+            sb.append(someRandomChars());
+            partsToAppend--;
+        }
+        // Occasionally make it a nested class:
+        if (r.nextBoolean()) {
+            sb.append('$');
+            sb.append(someRandomChars());
+        }
+        return DotName.createSimple(sb.toString());
+    }
+
+    private static DotName createRandomComponentised() {
+        int partsToAppend = r.nextInt(5) + 1;
+        DotName current = null;
+        while (partsToAppend > 0) {
+            current = DotName.createComponentized(current, someRandomChars());
+            partsToAppend--;
+        }
+        // Occasionally make it a nested class:
+        if (r.nextBoolean()) {
+            current = DotName.createComponentized(current, someRandomChars(), true);
+        }
+        return current;
+    }
+
+    private static String someRandomChars() {
+        final int chars = r.nextInt(10) + 1;
+        final char[] buf = new char[chars];
+        // Good enough for our purposes
+        final char[] validOptions = new char[] { 'a', 'b', 'c', 'd', 'e', '$' };
+        for (int i = 0; i < chars; i++) {
+            buf[i] = validOptions[r.nextInt(validOptions.length)];
+        }
+        return new String(buf);
+    }
+
+    private static DotName structuralCopy(DotName name) {
+        if (!name.isComponentized()) {
+            return DotName.createSimple(name.local());
+        }
+
+        DotName prefix = name.prefix();
+        if (prefix == null) {
+            return DotName.createComponentized(null, name.local(), name.isInner());
+        }
+        return DotName.createComponentized(structuralCopy(prefix), name.local(), name.isInner());
+    }
+
     @Test
     public void testIsComponentized() {
         assertTrue(DotName.createComponentized(DotName.createComponentized(null, "jboss"), "Foo").isComponentized());
@@ -125,10 +203,10 @@ public class DotNameTestCase {
             c.add(createRandomDotName());
         }
         // Throw in a special case:
-        c.add(DotName.createSimple("a"));
-        // "a" being the simplest case it must come first:
-        assertEquals("a", c.strings.iterator().next());
-        assertEquals("a", c.dotnames.iterator().next().toString());
+        c.add(DotName.createSimple("$"));
+        // "$" being the simplest case it must come first:
+        assertEquals("$", c.strings.iterator().next());
+        assertEquals("$", c.dotnames.iterator().next().toString());
         // Simple way to verify the comparator of DotName implements
         c.verifyAll();
     }
@@ -178,26 +256,6 @@ public class DotNameTestCase {
         definitelyNotEquals(simple_shorter, simple_perfectmatch);
     }
 
-    private void definitelyEquals(DotName a, DotName b) {
-        assertEquals(a, b);
-        assertEquals(b, a);
-        assertEquals(a.compareTo(b), 0);
-        assertEquals(b.compareTo(a), 0);
-        sameHashCode(a, b);
-    }
-
-    private static void sameHashCode(DotName a, DotName b) {
-        assertEquals(a.hashCode(), b.hashCode());
-    }
-
-    private void definitelyNotEquals(DotName a, DotName b) {
-        assertFalse(a.equals(b), "should not be equals");
-        assertFalse(b.equals(a), "should not be equals");
-        assertFalse(a.compareTo(b) == 0);
-        assertFalse(b.compareTo(a) == 0);
-        assertEquals(b.compareTo(a), -1 * a.compareTo(b));
-    }
-
     @Test
     public void testSpecialCase() {
         // This specific sequence was highlighting a bug
@@ -223,7 +281,7 @@ public class DotNameTestCase {
         DotName testName = DotName.createComponentized(indexerTestCase, Test$.class.getSimpleName(), true);
 
         Index index = Index.of(Test$.class);
-        assertEquals(testName, index.getKnownClasses().iterator().next().name());
+        definitelyEquals(testName, index.getKnownClasses().iterator().next().name());
         assertNotNull(index.getClassByName(DotName.createSimple(Test$.class.getName())));
         assertNotNull(index.getClassByName(testName));
     }
@@ -239,7 +297,7 @@ public class DotNameTestCase {
         DotName testName = DotName.createComponentized(test, $LeadingDelimiter.class.getSimpleName());
 
         Index index = Index.of($LeadingDelimiter.class);
-        assertEquals(testName, index.getKnownClasses().iterator().next().name());
+        definitelyEquals(testName, index.getKnownClasses().iterator().next().name());
         assertNotNull(index.getClassByName(DotName.createSimple($LeadingDelimiter.class.getName())));
         assertNotNull(index.getClassByName(testName));
     }
@@ -253,56 +311,78 @@ public class DotNameTestCase {
 
         Index index = Index.of($delimiters$.test.$SurroundedByDelimiters$.class);
         DotName indexedName = index.getKnownClasses().iterator().next().name();
-        assertEquals(testName, indexedName);
+        definitelyEquals(testName, indexedName);
         assertNotNull(index.getClassByName(testNameSimple));
         assertNotNull(index.getClassByName(testName));
         assertEquals("$delimiters$.test.$SurroundedByDelimiters$", indexedName.toString());
     }
 
-    private static DotName createRandomDotName() {
-        return r.nextBoolean() ? createRandomComponentised() : createRandomSimple();
+    @Test
+    public void componentizedEquals() {
+        for (int i = 0; i < 100; i++) {
+            DotName a = createRandomComponentised();
+            DotName b = structuralCopy(a);
+
+            definitelyEquals(a, b);
+        }
+
+        for (int i = 0; i < 100; i++) {
+            DotName a = createRandomComponentised();
+            DotName b = createRandomComponentised();
+
+            if (!a.toString().equals(b.toString())) {
+                definitelyNotEquals(a, b);
+            }
+        }
     }
 
-    private static DotName createRandomSimple() {
-        int partsToAppend = r.nextInt(5);
-        final StringBuilder sb = new StringBuilder();
-        sb.append(someRandomChars());
-        while (partsToAppend > 0) {
-            sb.append('.');
-            sb.append(someRandomChars());
-            partsToAppend--;
+    @Test
+    public void scalaAnonfunCurriedCase1() {
+        DotName a;
+        DotName b;
+
+        {
+            DotName scala = DotName.createComponentized(null, "scala", false);
+            DotName function2 = DotName.createComponentized(scala, "Function2$", false);
+            DotName anonfun = DotName.createComponentized(function2, "anonfun", true);
+            DotName curried = DotName.createComponentized(anonfun, "curried", true);
+            DotName one = DotName.createComponentized(curried, "1", true);
+            a = one;
         }
-        // Occasionally make it a nested class:
-        if (r.nextBoolean()) {
-            sb.append('$');
-            sb.append(someRandomChars());
+
+        {
+            DotName scala = DotName.createComponentized(null, "scala", false);
+            DotName function2 = DotName.createComponentized(scala, "Function2", false);
+            DotName anonfunCurriedOne = DotName.createComponentized(function2, "$anonfun$curried$1", true);
+            b = anonfunCurriedOne;
         }
-        return DotName.createSimple(sb.toString());
+
+        definitelyEquals(a, b);
     }
 
-    private static String someRandomChars() {
-        final int chars = r.nextInt(10) + 1;
-        final char[] buf = new char[chars];
-        // Good enough for our purposes, and want to avoid symbols:
-        final char[] validOptions = new char[] { 'a', 'b', 'c', 'd', 'e' };
-        for (int i = 0; i < chars; i++) {
-            buf[i] = validOptions[r.nextInt(validOptions.length)];
-        }
-        return new String(buf);
-    }
+    @Test
+    public void scalaAnonfunCurriedCase2() {
+        DotName a;
+        DotName b;
 
-    private static DotName createRandomComponentised() {
-        int partsToAppend = r.nextInt(5) + 1;
-        DotName current = null;
-        while (partsToAppend > 0) {
-            current = DotName.createComponentized(current, someRandomChars());
-            partsToAppend--;
+        {
+            DotName scala = DotName.createComponentized(null, "scala", false);
+            DotName function2 = DotName.createComponentized(scala, "Function2$", false);
+            DotName anonfun = DotName.createComponentized(function2, "anonfun", true);
+            DotName curried = DotName.createComponentized(anonfun, "curried", true);
+            DotName one = DotName.createComponentized(curried, "1", true);
+            a = one;
         }
-        // Occasionally make it a nested class:
-        if (r.nextBoolean()) {
-            current = DotName.createComponentized(current, someRandomChars(), true);
-        }
-        return current;
-    }
 
+        {
+            DotName scala = DotName.createComponentized(null, "scala", false);
+            DotName function2 = DotName.createComponentized(scala, "Function2", false);
+            DotName anonfun = DotName.createComponentized(function2, "$anonfun", true);
+            DotName curried = DotName.createComponentized(anonfun, "curried", true);
+            DotName one = DotName.createComponentized(curried, "1", true);
+            b = one;
+        }
+
+        definitelyEquals(a, b);
+    }
 }

